### PR TITLE
feat: even more precision for bash steps in github-env

### PR DESCRIPTION
**Pull request by ubiratansoares**
A follow-up for #197 

Here we leverage [tree-sitter-bash](https://crates.io/crates/tree-sitter-bash) rather than regexes to match usages of `$GITHUB_ENV` within a Bash script body. 

I took a simple approach, getting the full expansion for a `file_redirect` or `pipeline` and matching inner contents. Perhaps I'm missing some corner cases, but at least we have a starting point to iterate over 🙂

I also added some additional test cases (comments) and prepared the implementation to match against other types of shells in upcoming PRs.

### CODE REVIEW

File: src/audit/github_env.rs [hunk 1]
```diff
@@ -2,28 +2,66 @@ use super::{audit_meta, WorkflowAudit};
...
 pub(crate) struct GitHubEnv;
 
 audit_meta!(GitHubEnv, "github-env", "dangerous use of GITHUB_ENV");
 
 impl GitHubEnv {
-    fn uses_github_environment(run_step_body: &str) -> bool {
-        GITHUB_ENV_WRITE_SHELL.is_match(run_step_body)
+    fn evaluate_github_environment_within_bash_script(script_body: &str) -> anyhow::Result<bool> {
+        let bash = tree_sitter_bash::LANGUAGE;
+        let mut parser = Parser::new();
```
Comment (woodruffw) [inline src/audit/github_env.rs:17]: Nit: creating the parser with each evaluation is probably slower than necessary; we could probably reuse `parser` by making it a property of `GitHubEnv` that gets loaded when the audit itself is loaded with `GitHubEnv::new` 🙂

File: src/audit/github_env.rs [hunk 2]
```diff
@@ -47,8 +85,9 @@ impl WorkflowAudit for GitHubEnv {
             return Ok(findings);
         }
 
-        if let StepBody::Run { run, .. } = &step.deref().body {
-            if Self::uses_github_environment(run) {
+        if let StepBody::Run { run, shell, .. } = &step.deref().body {
+            let interpreter = shell.clone().unwrap_or("bash".into());
```
Comment (woodruffw) [inline src/audit/github_env.rs:89]: This is unfortunately not guaranteed to be the right default, since Windows runners default to `pwsh`. I think what we need to do is look at the surrounding `NormalJob::runs_on` and maybe add some kind of `RunsOn::default_shell()` helper. I'll look into that a bit.

(Nonblocker, since this is still improving on the precision of the current approach 🙂)

File: src/audit/github_env.rs [hunk 3]
```diff
@@ -72,7 +111,7 @@ mod tests {
     use crate::audit::github_env::GitHubEnv;
 
     #[test]
-    fn test_shell_patterns() {
+    fn test_exploitable_bash_patterns() {
```
Comment (woodruffw) [inline src/audit/github_env.rs:114]: I think we can fold these two test functions into one, with `(case, expected: bool)` as the parametrization 🙂

File: src/audit/github_env.rs (hunk 2 above)
Comment (woodruffw) [reply src/audit/github_env.rs:89]: Actually, it's a bit more annoying than even that: we need to determine the step's shell via the following precedence:

1. [defaults.run.shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#defaultsrunshell)
2. [jobs.<job_id>.defaults.run.shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_iddefaultsrunshell)
3. [jobs.<job_id>.steps[*].shell](https://docs.github.com/en/actions/writing-workflows/workflow-syntax-for-github-actions#jobsjob_idstepsshell)
4. If none of the above are set, use the default shell implied by the runner from `runs-on` (which might be indeterminate in the self-hosted runner case)

Comment (ubiratansoares) [reply src/audit/github_env.rs:89]: @woodruffw I totally missed `pwsh` on Windows as the default, I never used Windows runners and I though that `bash` was the default (via WLS). Another day, another learning I guess 🙂

Comment (woodruffw) [reply src/audit/github_env.rs:89]: I forget about too 🙂 -- the changes in #213 should enable this, but I'll rebase off of your work so that it's not a blocker here.

**Comment by ubiratansoares** [comment]
@woodruffw I lack the energy to work on anything else today, I'll resume the work on this PR tomorrow, addressing all your suggestions, and perhaps making the most of #213 🙂

Meanwhile, thanks again for the great review (as usual), it's a pleasure contributing to and learning from this project ❤️

**Comment by woodruffw** [comment]
Thank you too! And there's no rush or urgency at all, it's all supposed to be open source fun 🙂

File: src/audit/github_env.rs (hunk 1 above)
Comment (ubiratansoares) [reply src/audit/github_env.rs]: Done [here](https://github.com/woodruffw/zizmor/pull/208/commits/bc0e914233b4ac5209822e04f25faee2df6b573f), along with other changes!

**Review by woodruffw — APPROVED** [review]
LGTM, thanks a ton @ubiratansoares!
